### PR TITLE
journalctl: several fixes and cleanups for --follow

### LINKED ISSUE #26746: `journalctl --follow --cursor-file` does not update cursor
### systemd version the issue has been seen with

252.5-2~bpo11+1

### Used distribution

Debian 11

### Linux kernel version used

5.10.0-20-amd64

### CPU architectures issue was seen on

x86_64

### Component

journalctl

### Expected behaviour you didn't see

`journalctl --follow --cursor-file` should update the cursor file, at least when it exits normally. (I will leave it up to the developers what "normally" means here, but there should be *some* way to terminate journalctl so that it updates the cursor file.)

Ideally, `journalctl --follow --cursor-file` should just regularly update the cursor file while it is running, like `systemd-journal-upload --save-state` does.

### Unexpected behaviour you saw

I could not find any way to meaningfully use `journalctl --follow` with a cursor. Neither `--cursor-file` nor `--show-cursor` write/print a new cursor.

### Steps to reproduce the problem

1. Run `journalctl --follow --cursor-file=some-cursor-file`
2. Write something new to the journal
3. Terminate journalctl through Ctrl+C, SIGTERM, etc.
4. Observe that the cursor-file has not changed

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
root@bullseye:~# cat cursor 
s=ef839fb626f647fc92fbc870754bfe7d;i=349;b=1cd8ecb4ee404b9f844f5a701ef6d746;m=148d349f;t=5f689259aa9d8;x=570045226c850ed0
root@bullseye:~# SYSTEMD_LOG_LEVEL=debug journalctl --cursor-file=cursor --follow
Considering root directory '/run/log/journal'.
Root directory /run/log/journal added.
Considering root directory '/var/log/journal'.
Root directory /var/log/journal added.
Considering directory '/var/log/journal/ad92a246eb3a495897f8bc9263f08004'.
Directory /var/log/journal/ad92a246eb3a495897f8bc9263f08004 added.
Journal effective settings seal=no keyed_hash=yes compress=yes compress_threshold_bytes=8B
File /var/log/journal/ad92a246eb3a495897f8bc9263f08004/user-1000.journal added.
Journal effective settings seal=no keyed_hash=no compress=yes compress_threshold_bytes=8B
File /var/log/journal/ad92a246eb3a495897f8bc9263f08004/system@ef839fb626f647fc92fbc870754bfe7d-0000000000000001-0005f6890e1aa6f8.journal added.
File /var/log/journal/ad92a246eb3a495897f8bc9263f08004/user-1000@2820e11569904164aef78f2b0e016bc1-00000000000002bb-0005f6890e7a175f.journal added.
Journal effective settings seal=no keyed_hash=yes compress=yes compress_threshold_bytes=8B
File /var/log/journal/ad92a246eb3a495897f8bc9263f08004/system.journal added.
Journal filter: _BOOT_ID=1cd8ecb4ee404b9f844f5a701ef6d746
Reiterating files to get inotify watches established.
Considering root directory '/run/log/journal'.
Considering root directory '/var/log/journal'.
Considering directory '/var/log/journal/ad92a246eb3a495897f8bc9263f08004'.
-- Journal begins at Fri 2023-03-10 10:16:31 UTC. --
Mar 10 10:23:45 bullseye vagrant[2898]: Hello World!
^C
root@bullseye:~# cat cursor 
s=ef839fb626f647fc92fbc870754bfe7d;i=349;b=1cd8ecb4ee404b9f844f5a701ef6d746;m=148d349f;t=5f689259aa9d8;x=570045226c850ed0
```

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #26746.

### CODE REVIEW

**Review by YHNdnzj — APPROVED** [review]
Not very familiar with `sd_event_loop`, so I'm leaving the green label to someone else :)

Superficially all looks good to me.

File: src/journal/journalctl.c [hunk 1]
```diff
@@ -2753,25 +2786,15 @@ static int run(int argc, char *argv[]) {
...
-                for (;;) {
-                        r = wait_for_change(j, poll_fd);
-                        if (r < 0)
-                                return r;
+                assert(poll_fd >= 0);
 
-                        r = show(&c);
-                        if (r < 0)
-                                return r;
+                r = setup_event(&c, poll_fd, &e);
```
Comment (keszybz) [inline src/journal/journalctl.c:2774]: > One minor(?) behavior change is that 'journalctl --follow' now exits with 0 on terminate e.g. with ctrl-C.

That's actually a regression. I'm not sure if anyone relies on this, but previously we were returning correct information but now we're suppressing it.

E.g. when I run this silly example in fish, the result looks strange:
```console
[0] $ tail -f /etc/fstab | build/journalctl -f | grep foobar
^C⏎
[SIGINT|0|SIGINT] $
```

**Review by keszybz — COMMENTED** [review]
Looks good, except for the one issue with the event loop. I don't know of anyone relying on the return value, but OTOH, this is a standard UNIX interface to return the proper reason of exit, and I don't think we want to suppress it. Other tools certainly don't.

**Comment by yuwata** [comment]
@keszybz Thank you for the review.
Now the received signal is re-raised after the cursor is updated.
PTAL.

**Comment by yuwata** [comment]
Added testcase, fixed one more issue. PTAL.

File: man/sd_journal_next.xml [hunk 2]
```diff
@@ -77,6 +84,16 @@
     <para>Similarly, <function>sd_journal_previous()</function> sets
     the read pointer back one entry.</para>
 
+    <para><function>sd_journal_step_one()</function> also moves the read pointer. If the current location
+    is the head of the journal, e.g. when this is called followed by
```
Comment (YHNdnzj) [inline man/sd_journal_next.xml]: when this is called after/following `sd_journal_seek_head`?

File: man/sd_journal_next.xml [hunk 3]
```diff
@@ -77,6 +84,16 @@
...
+    <function>sd_journal_seek_head()</function>, then this is equivalent to
+    <function>sd_journal_next()</function>, and the argument <varname>advanced</varname> will be ignored.
+    Similary, if the current location is the tail of the journal, e.g. when this is called followed by
```
Comment (YHNdnzj) [inline man/sd_journal_next.xml]: same here

**Review by YHNdnzj — APPROVED** [review]
The code looks good to me, but I'm not sure about the comments. IIUC you actually mean _following_ where you talked about _followed by_?

**Comment by yuwata** [comment]
> The code looks good to me, but I'm not sure about the comments. IIUC you actually mean _following_ where you talked about _followed by_?

Yeah, exactly. Thank you!

**Pull request merged**